README.txt
Moodle Direct Release Notes
------------------------------------------------------------------------------------
Date:       2015-Aug-18
Release:    v2015030303

- Support for Moodle 2.9 including rewriting of submission workflow.
- API URL changed for UK accounts.
- Fixes:
    - Due date change no longer causes calendar event error.
    - Missing institutional check field added to Moodle database.

####################################################################################
Date:       2015-Jun-30
Release:    v2015030302

- Increase submission limit to Turnitin to 40Mb for newly created classes.
- Updated Catalan language pack.
- Log entry added for Assignment Resubmission.
- Categories link added to inbox bread crumb.
- Requirement paths consolidated.
- Fixes:
    - Bug with events table insertion in 2.5.
    - Plugin now works with PHP 5.6.6.
    - Sorting by submitted date in submission inbox.
    - Styles link corrected on index.php.
    - Settings warning notice due to jquery inclusion no longer appears.
    - jQuery require function in settings.
    - Include jQuery in page header on unlink users page.
    - Modulename lang string now a fixed value rather than relying on other string value.

####################################################################################
Date:       2014-Sept-08
Release:    v2015030301

- Fixes:
    > When enrolling all students, check student submit capability

####################################################################################
Date:       2014-Oct-31
Release:    v2014103101

- Added Czech language pack.

####################################################################################
Date:       2014-Sept-08
Release:    v2013111404

- Fixes:
    > Removed incorrect XML in install.xml
    > Changed jQuery include method to use Moodle jQuery where possible
    > 'Anonymous marking enabled' button is disabled if submission was not successful
    > Assignment type is required when creating an assignment, cannot be set to blank
    > Changed logging to use Moodle events where available
    > Fixed many depricated function warnings since the Moodle 2.7 release
    > Changed suggested API URL to api.turnitin.com
    > Added default value for $params in turnitintool_delete_records_select

####################################################################################
Date:       2014-Apr-04
Release:    v2013111403

- Anonymous marking option is locked once a submission is made to any assignment part
- Upgraded jQuery to 1.11.0
- Added cURL CA cert link to INSTALL.html
- Fixes:
    > Improved appearance of errors that appear during refresh submissions
    > Inbox sort on date now works with UK date format dd/mm/yy as well as mm/dd/yy
    > Fix for generic API errors being returned when using non-English language
    > Non-OR capable submissions now show "--" for similarity in assignment inbox
      rather than "0%"
    > Added file check to identify submissions that are missing from disk/corrupt in
      Moodle and remove the associated submission

####################################################################################
Date:       2014-Feb-26
Release:    v2013111402

- Fixes for apostrophes in Moodle 1.9

####################################################################################
Date:       2013-Nov-14
Release:    v2013111401

- In anonymous marking assignments, inbox displays in part view until all
  parts have passed the post date.
- Namespaced Turnitin JavaScript to avoid overwriting by other jQuery includes
- Polish and Russian strings added to language packs
- Help icons added for Turnitin advanced options
- Resubmission is now possible when reports generated on due date
- Fixes:
    > Submission inbox catches errors, no longer refreshes indefinitely
    > Refresh inbox row button now updates gradebook
    > Various fixes to SQL queries for different database types
    > Fixed default grade score
    > Fixed error when writing to logs

####################################################################################
Date:       2012-Dec-04
Release:    v2012120401

- Refactored Submission inbox views, jQuery filtering implemented to reduce database requests
- Refactored 'Enroll All Students' and 'Refresh Submissions' to use ajax
- Optimised database queries on inbox views and refresh submissions calls
- Added more granular logging, now logging submission add and delete, assignment add, delete and update
- Added a submission event handler (for Moodle 2+ only)

####################################################################################
Date:       2012-Nov-04
Release:    v2012110401

- Moodle Direct now requires PHP 5.0 server environments and above
- Added performance improvements, specifically in-box refresh on large assignments
- Added time-zone synchronisation improvements

####################################################################################
Date:       2012-Sept-24
Release:    v2012092401

- Added support for Translated Matching
- Update icons with current Turnitin icons, update icons to allow multi version support
- Added a config level mod_turnitintool component file browser when using Moodle 2.0+ for mod_turnitintool
- Re-factored Unlink / Relink users page to paginate user data and reduce memory usage

####################################################################################
Date:       2011-Aug-18
Release:    v2011081801

- Refactored Back up and restore to allow duplication of TII classes and assignments
- Added erater / ETS support
- Added additional email notification options in the admin config screen

####################################################################################
Date:       2011-July-29
Release:    v2011072901

- Added support Bulk Download of Submissions in PDF and Original format
- Added feature to download grade report XLS spreadsheet
- Added Multi tutor management screen

####################################################################################
Date:       2010-Nov-19
Release:    v2010111901

- Added support for Moodle groups

####################################################################################
Date:       2010-Oct-26
Release:    v2010102601

- Added pagination to the inbox
- Updated database fields and tables for Oracle support
- Added exclude small matches global assignment setting
- Added support for multi language api calls
- Added French (fr) language string file
- Fixes:
    > Fixed issue where non enrolled students were not displayed in the tutor inbox view
    > Fixed issue where user's resubmissions where incorrectly tagged as anonymous
    > Fixed issue with incorrect / incomplete ordering of anonymous inbox

####################################################################################
Date:       2010-Sept-01
Release:    v2010090101

- Added various changes to add compatibility for Moodle 2.0
    > Refactored table output to support both Moodle 1.9 - 2.0
    > Updated language pack, incorporated help into standard language strings
    > Updated Javascript and CSS file functionality
    > Added Moodle 2.0 Back Up and Restore
    > Changed file storage to use Moodle 2.0 file storage where available
    > Moved images to 'pix' directory instead of 'images'
- Added Backup and Restore for Moodle 1.9

####################################################################################
Date:       2010-June-19
Release:    v2010061901

- Added additional diagnostic logging
- Added Authenticated Proxy support

####################################################################################
Date:       2010-June-12
Release:    v2010061201

- Refactored Inbox SQL queries

####################################################################################
Date:       2010-June-2
Release:    v2010060201

- Added support for UTF-8 intepretation of API return data

####################################################################################
Date:       2010-April-23
Release:    v2010042301

- Removed redundant assignment synching cron functionality
- Now allows resubmission to the same paper ID

####################################################################################
Date:       2010-April-06
Release:    v2010040601
 
- Provides seamless integration into Turnitin using Moodle workflow 
- Uses an activity module so that we can update Turnitin independently of Moodle 
- Uses real Turnitin accounts to allow users to log directly in to Turnitin (should they need to) 
- Uses a 'pull' approach to information and has no 'call-backs' to the local VLE 
- Will run behind a fire wall 
- Will handle multi-part assignments (one assignment many files) 
- Sends GradeMark marking information to the Moodle GradeBook 
- Grades are not released until the due date 
- Course recycle will correctly copy forward Turnitin information 
- Tutors can load work on behalf of students 
- Turnitin classes can only have one owner. The class owner is set to the person that created the course in Moodle. Only the class owner will be able to see the assignments when logging in to Native Turnitin. However you can change the class owner from within Moodle if you are an instructor. 

Current Limitations
- Does not support Oracle as the Moodle database
- Will not update information changed in Turnitin by users in native Turnitin 
- Will not work with assignments created under the framed-in API 
- No support for revision assignment, master classes, GradeMark analytics, translations, Zip file upload, PeerMark, QuickSubmit


ChangeLog.txt
2015-08-18 : Support for Moodle 2.9 including rewriting of submission workflow.
2015-08-18 : API URL changed for UK accounts.
2015-08-18 : Due date change no longer causes calendar event error.
2015-08-18 : Missing institutional check field added to Moodle database.
2015-06-30 : Increase submission limit to Turnitin to 40Mb for newly created classes.
2015-06-30 : Updated Catalan language pack.
2015-06-30 : Log entry added for Assignment Resubmission.
2015-06-30 : Categories link added to inbox bread crumb.
2015-06-30 : Requirement paths consolidated.
2015-06-30 : Fixes:
2015-06-30 : Bug with events table insertion in 2.5.
2015-06-30 : Plugin now works with PHP 5.6.6.
2015-06-30 : Sorting by submitted date in submission inbox.
2015-06-30 : Styles link corrected on index.php.
2015-06-30 : Settings warning notice due to jquery inclusion no longer appears.
2015-06-30 : jQuery require function in settings.
2015-06-30 : Include jQuery in page header on unlink users page.
2015-06-30 : Modulename lang string now a fixed value rather than relying on other string value.
2015-03-03 : When enrolling all students, check student submit capability
2014-10-31 : Added cs language pack
2014-10-02 : Added default value for $params in turnitintool_delete_records_select
2014-09-08 : Removed incorrect XML in install.xml
2014-09-08 : Changed jQuery include method to use Moodle jQuery where possible
2014-09-08 : 'Anonymous marking enabled' button is disabled if submission was not successful
2014-09-08 : Assignment type is required when creating an assignment, cannot be set to blank
2014-09-08 : Changed logging to use Moodle events where available
2014-09-08 : Fixed many depricated function warnings since the Moodle 2.7 release
2014-09-08 : Changed suggested API URL to api.turnitin.com
2014-04-04 : Anonymous marking option is locked once a submission is made to any assignment part
2014-04-04 : Upgraded jQuery to 1.11.0
2014-04-04 : Added cURL CA cert link to INSTALL.html
2014-04-04 : Improved appearance of errors that appear during refresh submissions
2014-04-04 : Inbox sort on date now works with UK date format dd/mm/yy as well as mm/dd/yy
2014-04-04 : Fix for generic API errors being returned when using non-English language
2014-04-04 : Non-OR capable submissions now show "--" for similarity in assignment inbox rather than "0%"
2014-04-04 : Added file check to identify submissions that are missing from disk/corrupt in Moodle and remove the associated submission
2014-02-26 : Changes to prevent apostrophe issues in Moodle 1.9
2013-11-14 : Fixed assingment inbox so that it will error out instead of refreshing indefinitely
2013-11-14 : Fixed user search in unlink/relink screen
2013-11-14 : Submission logic fixed to allow resubmission until due date when reports generated on due date
2013-11-14 : Assignment inbox stays in part view until all parts pass post date
2013-11-14 : Changed the write to log function as it was causing errors in some environments
2013-11-14 : Fixed various issues on postgres databases
2013-11-14 : Added Russian and Polish strings to language packs
2013-11-14 : Fixed refresh inbox row button
2013-11-14 : Turnitin JavaScript namespaced to avoid overwriting by jQuery
2013-05-20 : Refresh individual inbox row now fetches Translated Matching similarity
2013-05-20 : Fixed issue where deleted parts were being flagged as duplicates after course reset
2013-05-20 : Added support for backing up / restoring activity logs
2013-05-20 : Added support for .wpd and .ps file types
2013-05-20 : Warning that grades will be overwritten now correctly disables after due date rather than post date
2013-05-20 : Fixed issue where user's were not able to resubmit for the option 'Generate reports on due date'
2013-05-20 : Fixed issue with Student data privacy auto generated last name
2013-05-20 : Fixed isse where submission modified dates
2013-05-20 : Fixed issue in database delete call when updating inbox and purging old records
2013-05-20 : Fixed issue where updating individual row set the submitted date to now()
2013-04-26 : Using conventional names for all classes, constants and method names outside classes should begin TURNITINTOOL_ or turnitintool_
2013-04-26 : Added support for quick name change in course home page where applicable to Moodle version
2013-04-26 : Added support for .odt and .hwp files
2013-04-26 : Fixed issue where old records that no longer exist in Turnitin were not being purged from inbox
2013-04-26 : Added ability to refresh individual rows, this will also purge duplicates if found
2013-04-26 : Fixed issue where db calls were called inside a loop on the submission form screen
2013-03-19 : Adding alternative method for adding curl certificate bundles
2013-03-19 : Fixed issue with using 'AS' in Oracle DB column aliases in JOIN clauses
2013-03-19 : Fixed regression where 0% OR was displayed as 'Pending'
2013-03-19 : Added missing support for groups in new inbox function
2013-03-05 : Added fix for differing profile language when loading javascript before login
2013-03-05 : Added some user checks to the javascript jump to page
2013-02-25 : Fixed a warning when refreshing inbox, which caused unnecessary database updates
2013-02-25 : Fixed a Moodel 1.9 backwards compatibility issue
2013-02-25 : Fixed some postgres related DB issues
2013-02-18 : Fixed issue where deleted anonymous marking parts were making export links disappear
2013-02-18 : Fixed issue where redirects were failing if the Moodle site was ran under a sub directory e.g. examplemoodle.org/moodle/
2013-02-18 : Added the submission title back into the anonymous marking in box rows
2013-02-18 : Fixed issue in postgres when accessing student My Submissions page
2012-12-04 : Moved 'Enroll All Students' actions into ajax, enroll all students can now happen while refresh submissions is running

view.php
<?php
/**
 * @package   turnitintool
 * @copyright 2012 Turnitin
 */

require_once(__DIR__."/../../config.php");
require_once(__DIR__."/lib.php");
require_once($CFG->libdir."/completionlib.php");
require_once($CFG->libdir."/formslib.php");
require_once($CFG->libdir."/form/text.php");
require_once($CFG->libdir."/form/datetimeselector.php");
require_once($CFG->libdir."/form/hidden.php");
require_once($CFG->libdir."/form/button.php");
require_once($CFG->libdir."/form/submit.php");
require_once(__DIR__."/version.php");

if (!turnitintool_check_config()) {
    turnitintool_print_error('configureerror','turnitintool',NULL,NULL,__FILE__,__LINE__);
    exit();
}
$viewpage='view.php';
$viewpage .= (isset($_REQUEST['id'])) ? '?id='.$_REQUEST['id'] : '';
$viewpage .= (isset($_REQUEST['do'])) ? '&do='.$_REQUEST['do'] : '';
turnitintool_activitylog($viewpage,"REQUEST");
turnitintool_activitylog("lib.php Loaded","REQUIRE_ONCE");
require_once($CFG->dirroot."/lib/uploadlib.php");
turnitintool_activitylog("uploadlib.php Loaded","REQUIRE_ONCE");

turnitintool_process_api_error();

$id = required_param('id', PARAM_INT); // Course Module ID, or
$a  = optional_param('a', 0, PARAM_INT);  // turnitintool ID

if ($id) {
    if (! $cm = get_coursemodule_from_id('turnitintool', $id)) {
        turnitintool_print_error("Course Module ID was incorrect");
    }

    if (! $course = turnitintool_get_record("course", "id", $cm->course)) {
        turnitintool_print_error("Course is misconfigured");
    }

    if (! $turnitintool = turnitintool_get_record("turnitintool", "id", $cm->instance)) {
        turnitintool_print_error("Course module is incorrect");
    }

} else {
    if (! $turnitintool = turnitintool_get_record("turnitintool", "id", $a)) {
        turnitintool_print_error("Course module is incorrect");
    }
    if (! $course = turnitintool_get_record("course", "id", $turnitintool->course)) {
        turnitintool_print_error("Course is misconfigured");
    }
    if (! $cm = get_coursemodule_from_instance("turnitintool", $turnitintool->id, $course->id)) {
        turnitintool_print_error("Course Module ID was incorrect");
    }
}

require_login($course->id);

if (isset($PAGE) AND is_callable(array($PAGE->requires, 'js'))) { // Are we using new moodle or old?
    if (!is_callable(array('page_requirements_manager', 'jquery'))) {
        $jsurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/scripts/jquery-1.11.0.min.js');
        $PAGE->requires->js($jsurl, true);
    } else {
        $PAGE->requires->jquery();
    }
    $jsurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/scripts/datatables.min.js');
    $PAGE->requires->js($jsurl);
    $jsurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/scripts/datatables.plugins.js');
    $PAGE->requires->js($jsurl);
    $jsurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/scripts/inboxtable.js');
    $PAGE->requires->js($jsurl);
    $jsurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/scripts/turnitintool.js');
    $PAGE->requires->js($jsurl,true);
    $cssurl = new moodle_url($CFG->wwwroot.'/mod/turnitintool/styles.css');
    $PAGE->requires->css($cssurl);
} else {
    require_js($CFG->wwwroot.'/mod/turnitintool/scripts/jquery-1.11.0.min.js');
    require_js($CFG->wwwroot.'/mod/turnitintool/scripts/turnitintool.js');
}
turnitintool_activitylog("turnitintool.js Loaded","REQUIRE_JS");

$param_jumppage=optional_param('jumppage',null,PARAM_CLEAN);
$param_userid=optional_param('userid',null,PARAM_CLEAN);
$param_post=optional_param('post',null,PARAM_CLEAN);
$param_delete=optional_param('delete',null,PARAM_CLEAN);
$param_update=optional_param('update',null,PARAM_CLEAN);
$param_do=optional_param('do',null,PARAM_CLEAN);
$param_enroll=optional_param('enroll',null,PARAM_CLEAN);
$param_owner=optional_param('owner',null,PARAM_CLEAN);
$param_anonid=optional_param('anonid',null,PARAM_CLEAN);
$param_updategrade=optional_param('updategrade',null,PARAM_CLEAN);
$param_up=optional_param('up',null,PARAM_CLEAN);
$param_submissiontype=optional_param('submissiontype',null,PARAM_CLEAN);
$param_submitted=optional_param('submitted',null,PARAM_CLEAN);
$param_delpart=optional_param('delpart',null,PARAM_CLEAN);
$param_unenrol=optional_param('unenrol',null,PARAM_CLEAN);
$param_enroltutor=optional_param('enroltutor',null,PARAM_CLEAN);
$param_s=optional_param('s',null,PARAM_CLEAN);
$param_ob=optional_param('ob',null,PARAM_CLEAN);
$param_sh=optional_param('sh',null,PARAM_CLEAN);
$param_fr=optional_param('fr',null,PARAM_CLEAN);
$param_export_data=optional_param('export_data',null,PARAM_CLEAN);
$param_objectid=optional_param('objectid',null,PARAM_CLEAN);
$param_partid=optional_param('partid',null,PARAM_CLEAN);
$param_utp=optional_param('utp',null,PARAM_CLEAN);
$param_enrollstudent=optional_param('enrollstudent',null,PARAM_CLEAN);
$param_reloadrow=optional_param('reloadrow',null,PARAM_CLEAN);

// Clean the post array so we can pass it into the functions below
$post = array();
foreach ( $_POST as $key => $value ) {
    // If 1.9 use clean_param to clean an array  / if 2.0 use clean_param_array
    if ( is_array( $value ) AND is_callable( 'clean_param_array' ) ) {
        $post[$key] = clean_param_array( $value, PARAM_CLEAN );
    } else {
        $post[$key] = clean_param( $value, PARAM_CLEAN );
    }
}

if (!is_null($param_jumppage)) {
    turnitintool_url_jumpto($param_userid,$param_jumppage,$turnitintool,$param_utp,$param_objectid,$param_partid,$param_export_data);
    exit();
}

if (!is_null($param_delete)) {
    if (!$submission = turnitintool_get_record('turnitintool_submissions','id',$param_delete)) {
        print_error('submissiongeterror','turnitintool');
        exit();
    }
    turnitintool_delete_submission($cm,$turnitintool,$USER->id,$submission);
    exit();
}

$redirectlink=$CFG->wwwroot.'/mod/turnitintool/view.php?id='.$cm->id;
$redirectlink.=(!is_null($param_do)) ? '&do='.$param_do : '&do=intro';
$redirectlink.=(!is_null($param_fr)) ? '&fr='.$param_fr : '';
$redirectlink.=(!is_null($param_sh)) ? '&sh='.$param_sh : '';
$redirectlink.=(!is_null($param_ob)) ? '&ob='.$param_ob : '';

if (!is_null($param_reloadrow)) {
    turnitintool_reloadinbox_row($cm,$turnitintool,$param_reloadrow);
    exit();
}

if (!is_null($param_update)) {
    $loaderbar = new turnitintool_loaderbarclass(2);
    turnitintool_update_all_report_scores($cm,$turnitintool,$param_update,$loaderbar);
    turnitintool_redirect($redirectlink);
    exit();
}

if (!is_null($param_enrollstudent)) {
    turnitintool_enroll_student($cm,$turnitintool,$param_enrollstudent);
}

if (!is_null($param_enroll)) {
    turnitintool_enroll_all_students($cm,$turnitintool);
    turnitintool_redirect($redirectlink);
    exit();
}

if (!is_null($param_do) AND $param_do=="tutors" AND !is_null($param_unenrol)) {
    $tutors=turnitintool_remove_tiitutor($cm,$turnitintool,$param_unenrol);
    if (!is_null($tutors->error)) {
        $notice=$tutors->error;
    } else {
        turnitintool_redirect($CFG->wwwroot.'/mod/turnitintool/view.php?id='.$cm->id.'&do=tutors');
        exit();
    }
}

if (!is_null($param_do) AND $param_do=="tutors" AND !is_null($param_enroltutor)) {
    $tutors=turnitintool_add_tiitutor($cm,$turnitintool,$param_enroltutor);
    if (!is_null($tutors->error)) {
        $notice=$tutors->error;
    } else {
        turnitintool_redirect($CFG->wwwroot.'/mod/turnitintool/view.php?id='.$cm->id.'&do=tutors');
        exit();
    }
}

if (!is_null($param_do) AND $param_do=="tutors" AND is_null($param_unenrol) AND is_null($param_enroltutor)) {
    $tutors=turnitintool_get_tiitutors($cm,$turnitintool);
    if (!is_null($tutors->error)) {
        $notice=$tutors->error;
    } else {
        $notice=null;
    }
}

if (!is_null($param_do) AND $param_do=="changeowner") {
    turnitintool_ownerprocess($cm,$turnitintool,$param_owner);
    turnitintool_redirect($CFG->wwwroot.'/mod/turnitintool/view.php?id='.$cm->id.'&do=tutors');
    exit();
}

if (!is_null($param_do) AND $turnitintool->autoupdates==1 AND $param_do=="allsubmissions" AND !is_null($param_anonid)) {
    $loaderbar = new turnitintool_loaderbarclass(3);
    turnitintool_revealuser($cm,$turnitintool,$post,$loaderbar);
    turnitintool_redirect($CFG->wwwroot.'/mod/turnitintool/view.php?id='.$cm->id.'&do=allsubmissions&update=1');
    exit();
}

if (!is_null($param_updategrade) OR isset($post['updategrade']) OR isset($post["updategrade_x"])) {
    turnitintool_update_form_grades($cm,$turnitintool,$post);
}

if (!is_null($param_up)) { // Manual Submission to Turnitin
    if (!$submission = turnitintool_get_record('turnitintool_submissions','id',$param_up)) {
        print_error('submissiongeterror','turnitintool');
        exit();
    }
    turnitintool_upload_submission($cm,$turnitintool,$submission);
    exit();
}

if (!is_null($param_submissiontype) AND $param_do=='submissions') {
    if ( (isset($param_userid)) && (has_capability('mod/turnitintool:grade', turnitintool_get_context('MODULE', $cm->id))) ) {
        $thisuserid=$param_userid;
    } else {
        $thisuserid=$USER->id;
    }

    if ($param_submissiontype==1) {
        if ($CFG->branch >= 29) {
            $notice = turnitintool_dofileupload_post_29($cm,$turnitintool,$thisuserid,$post);
        } else {
            $notice = turnitintool_dofileupload_pre_29($cm,$turnitintool,$thisuserid,$post);
        }
    } else if ($param_submissiontype==2) {
        $notice=turnitintool_dotextsubmission($cm,$turnitintool,$thisuserid,$post);
    }
    if ($turnitintool->autosubmission AND !empty($notice["subid"])) {
        if (!$submission = turnitintool_get_record('turnitintool_submissions','id',$notice["subid"])) {
            print_error('submissiongeterror','turnitintool');
            exit();
        }
        turnitintool_upload_submission($cm,$turnitintool,$submission);
        exit();
    }
}

if (!is_null($param_submitted) AND $param_do=='intro') {
    $notice=turnitintool_update_partnames($cm,$turnitintool,$post);
}

if (!is_null($param_delpart) AND $param_do=='intro') {
    $notice=turnitintool_delete_part($cm,$turnitintool,$param_delpart);
}

if (!is_null($param_submitted) AND $param_do=='notes') {
    $notice=turnitintool_process_notes($cm,$turnitintool,$param_s,$post);
}

if (!is_null($param_submitted) AND $param_do=='options') {
    $notice=turnitintool_process_options($cm,$turnitintool,$post);
}

turnitintool_add_to_log($course->id, "view turnitintool", "view.php?id=$cm->id", "User viewed assignment '$turnitintool->name'", "$cm->id");

// Enable activity completion on page view.
if (property_exists($CFG, 'branch') && $CFG->branch >= 20) {
    $completion = new completion_info($course);
    $completion->set_module_viewed($cm);
}

/// Print the page header
$strturnitintools = get_string("modulenameplural", "turnitintool");
$strturnitintool  = get_string("modulename", "turnitintool");

if(is_object($PAGE) && @is_callable(array($PAGE->navbar, 'add'))) {
    $navigation = '';
} elseif (!is_callable('build_navigation')) {
    $navigation = array(
            array('title' => $course->shortname, 'url' => $CFG->wwwroot."/course/view.php?id=$course->id", 'type' => 'course'),
            array('title' => $strturnitintools, 'url' => $CFG->wwwroot."/mod/turnitintool/index.php?id=$course->id", 'type' => 'activity'),
            array('title' => format_string($turnitintool->name), 'url' => '', 'type' => 'activityinstance')
    );
} else {
    $navigation = build_navigation('',$cm);
}

// Do not use navbar in 2.7+
if ((property_exists($CFG, 'branch') AND ($CFG->branch < 27)) || (!property_exists($CFG, 'branch'))) {
turnitintool_header($cm,
        $course,
        $_SERVER["REQUEST_URI"],
        $turnitintool->name,
        $SITE->fullname,
        $navigation,
        "",
        "",
        true,
        update_module_button($cm->id, $course->id, $strturnitintool),
        navmenu($course)
);
} else {
turnitintool_header($cm,
        $course,
        $_SERVER["REQUEST_URI"],
        $turnitintool->name,
        $SITE->fullname,
        $navigation,
        "",
        "",
        true,
        update_module_button($cm->id, $course->id, $strturnitintool)
);
}

/// Check to see if groups are being used and abstract for 1.8 if neccessary
if (!is_callable('groups_get_activity_group')) {
    $changegroup = optional_param('group', -1, PARAM_INT);
    $cm->currentgroup=get_and_set_current_group($course, $cm->groupmode, $changegroup);
    setup_and_print_groups($course, $cm->groupmode, $redirectlink);
} else {
    $groupmode = groups_get_activity_groupmode($cm);
    if ($groupmode) {
        groups_get_activity_group($cm, true);
        groups_print_activity_menu($cm, $redirectlink);
    }
}

// Print the main part of the page
echo '<div id="turnitintool_style">';

if (!is_null($param_do)) {
    $do=$param_do;
} else {
    $do='intro';
}

// $do=ACTION
// $do=submissions >>> Student Submission Page
// $do=intro >>> Turnitin Assignment Intro Page
// $do=allsubmissions >>> Tutor View All Submissions
// $do=bulkupload >>> Tutor Bulk Upload Student Submissions
// $do=viewtext >>> View Student Text Submission
// $do=submissiondetails >>> View Submission Details

$studentdos=array('submissions','intro','viewtext','submissiondetails','notes');
$graderdos=array('allsubmissions','options','changeowner','tutors');

// If an unrecognised DO request produce error
if (!in_array($do,$studentdos) AND !in_array($do,$graderdos)) {
    turnitintool_print_error('dorequesterror','turnitintool');
    exit();
} else if (!has_capability('mod/turnitintool:grade', turnitintool_get_context('MODULE', $cm->id)) AND in_array($do,$graderdos)) {
    turnitintool_print_error('permissiondeniederror','turnitintool');
    exit();
}

echo '<br />';
turnitintool_draw_menu($cm,$do);

if ($do=='intro') {
    if (isset($notice['error'])) {
        turnitintool_box_start('generalbox boxwidthwide boxaligncenter error', 'errorbox');
        echo $notice['message'];
        turnitintool_box_end();
    } else {
        $notice=NULL;
    }
    // Update the GradeBook to make sure the grade stays 'hidden' until and wasn't revealed by modedit
    turnitintool_grade_item_update( $turnitintool );
    echo turnitintool_duplicatewarning($cm,$turnitintool);
    echo turnitintool_introduction($cm,$turnitintool,$notice);
}

if ($do=='submissions') {
    if ( !has_capability('mod/turnitintool:grade', turnitintool_get_context('MODULE', $cm->id))
         AND !has_capability('mod/turnitintool:submit', turnitintool_get_context('MODULE', $cm->id))) {
        turnitintool_print_error('permissiondeniederror','turnitintool');
        exit();
    } else {
        echo turnitintool_view_student_submissions($cm,$turnitintool);
        if (isset($notice["error"]) && ($CFG->branch < 29)) {
            turnitintool_box_start('generalbox boxwidthwide boxaligncenter error', 'errorbox');
            echo $notice["error"];
            turnitintool_box_end();
        }
        echo turnitintool_view_submission_form($cm,$turnitintool);
    }
}

if ($do=='allsubmissions') {
    if (!empty($notice)) {
        turnitintool_box_start('generalbox boxwidthwide boxaligncenter error', 'errorbox');
        echo $notice;
        turnitintool_box_end();
    }
    if (isset($param_ob)) {
        $ob=$param_ob;
    } else {
        $ob=1;
    }
    echo turnitintool_view_all_submissions($cm,$turnitintool,$ob);
}

if ($do=='notes') {
    echo turnitintool_view_notes($cm,$turnitintool,$param_s,$post);
    if (isset($notice['error'])) {
        turnitintool_box_start('generalbox boxwidthwide boxaligncenter error', 'errorbox');
        echo $notice['message'];
        turnitintool_box_end();
    } else {
        $notice=NULL;
    }
    echo turnitintool_addedit_notes($cm,$turnitintool,$param_s,$post,$notice);

}

if ($do=='options') {
    if (!empty($notice)) {
        turnitintool_box_start('generalbox boxwidthwide boxaligncenter', 'general');
        echo $notice;
        turnitintool_box_end();
    }
    echo turnitintool_view_options($cm,$turnitintool);
}

if ($do=='tutors') {
    if (!is_null($notice)) {
        turnitintool_box_start('generalbox boxwidthwide boxaligncenter error', 'errorbox');
        echo $notice;
        turnitintool_box_end();
    } else {
        echo turnitintool_view_tiitutors($cm,$turnitintool,$tutors);
    }
}

// Finish the page
echo '</div>';

if (isset($PAGE) AND @is_callable(array($PAGE->requires, 'js'))) { // Are we using new moodle or old?
    // We already added the Moodle 2.0+ stuff
} else {
    // These need to go to the botton here to avoid conflicts
    require_js($CFG->wwwroot.'/mod/turnitintool/scripts/datatables.min.js');
    require_js($CFG->wwwroot.'/mod/turnitintool/scripts/datatables.plugins.js');
    require_js($CFG->wwwroot.'/mod/turnitintool/scripts/inboxtable.js');
}

turnitintool_footer($course);
$parts=turnitintool_get_records('turnitintool_parts','turnitintoolid',$turnitintool->id);
$parts_string="(";
foreach ($parts as $part) {
    $parts_string.=($parts_string!="(") ? " | " : "";
    $parts_string.= $part->partname.': '.$part->tiiassignid;
}
$parts_string.=")";
echo '<!-- Turnitin Moodle Direct Version: '.$module->version.' - '.$parts_string.' -->';

/* ?> */